Masterbar: Add shopping cart button in app for use in wp-admin

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

https://github.com/Automattic/wp-calypso/pull/59470 added the `mini-cart` package and https://github.com/Automattic/wp-calypso/pull/58467 used that package to display a popup version of the shopping cart on all single-site pages in calypso (if the cart is not empty). 

This PR creates an app within the calypso monorepo to allow embedding the popup shopping cart outside of calypso, specifically in the wp-admin masterbar. This will unify the masterbar across page redirects between wp-admin and calypso where the masterbar cart might otherwise seem to disappear due to nav unification having made everything else look the same.

![masterbar-cart](https://user-images.githubusercontent.com/2036909/132778713-06977670-e7c2-46fb-8e8b-ce69d4a440c5.png)

#### Testing instructions

- `cd apps/masterbar-cart`
- `yarn && yarn start`
- Visit http://localhost:3000/?site=YOUR-SITE-ID-HERE (the number, not the name).

a8ci18n commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -287,6 +289,16 @@ class MasterbarLoggedIn extends React.Component {
+							require="./masterbar-cart"
+							placeholder={ null }
+							className="masterbar__item-cart"
+							tooltip={ translate( 'View my Shopping Cart' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **34** times:
`translate( 'My Shopping Cart' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/masterbar-cart)

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx:
```diff
@@ -271,15 +273,19 @@ export function WPNonProductLineItem( {
 					>
-						<DeleteIcon uniqueID={ deleteButtonId } product={ label } />
+						{ isDeleteButtonText ? (
+							translate( 'Remove from cart' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **32** times:
`translate( 'Remove from Cart' )` *ES Score: 9*

michaeldcain commented:
@sirbrillig I just pushed a few commits to do the following:

- rebase from trunk to pull in any of the fun shopping cart querying you've been working on
- moved from inline styles to an included stylesheet to take advantage of style linting and RTL automation
- set a default width to prevent the popover from jumping around
- make the delete button a text link via new prop
- add a site title (we might want to limit this to only users with multiple sites)

michaeldcain commented:
Oh, screenshot for record:

<img width="591" alt="Screen Shot 2021-08-16 at 1 38 48 PM" src="https://user-images.githubusercontent.com/942359/129606428-0357e5c6-0dd5-4db4-abb2-ac41e0e22829.png">

michaeldcain commented:
A cart with more items:

<img width="1552" alt="Screen Shot 2021-08-16 at 1 56 30 PM" src="https://user-images.githubusercontent.com/942359/129610789-888d472a-5099-4f64-931d-f6ce32c5a657.png">

@sirbrillig should we remove the popover carts from the domains and plans pages in this PR?

michaeldcain commented:
This doesn't currently work on wp-admin pages, which is particularly messy with the unified nav:
![unified-nav-cart](https://user-images.githubusercontent.com/942359/129620814-2ee2b699-0769-4ea1-b9fd-046adff13353.gif)

sirbrillig commented:
> @sirbrillig should we remove the popover carts from the domains and plans pages in this PR?

They should be removed for certain. I'm not sure if it's best to do so in this PR or a separate one...

sirbrillig commented:
In order for this to work outside of calypso (ie: in wp-admin) we'll need to make sure all its dependencies are external or can be easily replaced. Here's the current internal dependencies for some discussion:

```js
import Popover from 'calypso/components/popover';
```

Popover is being moved to external in https://github.com/Automattic/wp-calypso/pull/55238

```js
import CalypsoShoppingCartProvider from 'calypso/my-sites/checkout/calypso-shopping-cart-provider';
```

As this is basically just a wrapper, it can be fairly easily recreated with `ShoppingCartProvider`, some mechanism to get the cart key (which is just the side ID so I don't think it will be hard in wp-admin), and a way to call the `/me/shopping-cart` endpoints (which we can do using `wpcomProxyRequest` and wpcom.js).

```js
import CartMessages from 'calypso/my-sites/checkout/cart/cart-messages';
```

This could be excluded as the popover cart doesn't need to display error messages all that often, but it could also be recreated in some way if desired.

```js
import { CheckoutSummaryTotal } from 'calypso/my-sites/checkout/composite-checkout/components/wp-checkout-order-summary';
import { WPOrderReviewLineItems } from 'calypso/my-sites/checkout/composite-checkout/components/wp-order-review-line-items';
```

These two are the biggest dependencies. The `@automattic/composite-checkout` package includes simplified versions of these components but we really want to use the ones in calypso if possible since they include all the product-specific formatting that we use. I will work on trying to move these components to `@automattic/wpcom-checkout`.

```js
import { errorNotice, infoNotice, successNotice } from 'calypso/state/notices/actions';
```

These are probably not really used by the popover cart, but could easily be replaced by some other sort of notices.

```js
import { getSelectedSite } from 'calypso/state/ui/selectors';
```

This is just used to get the site ID; there are other ways to do that without the Redux tree.

```js
import MasterbarItem from './item';
```

This is probably the biggest thing we'd have to recreate; it's the mechanism which holds the menubar item.

jsnajdr commented:
> Popover is being moved to external in #55238

The `Popover` move to `@automattic/components` has been merged.

jsnajdr commented:
> `import { errorNotice, infoNotice, successNotice }`

These notice actions are used to construct callbacks like `showErrorMessage` passed as props to `CheckoutProvider`. It should be possible to replace the callback with another implementation, showing `wp-admin` notices.

sirbrillig commented:
> WPOrderReviewLineItems

Looking more closely at this, the dependencies for line items are as follows.

```js
import { getSublabel, getLabel } from '../lib/translate-cart';
```

These are going to be moved to `@automattic/wpcom-checkout` in https://github.com/Automattic/wp-calypso/pull/55734

```js
import { currentUserHasFlag, getCurrentUser } from 'calypso/state/current-user/selectors';
```

Both of these are used only to determine if the user is eligible to buy a domain without a plan, and the only thing affected is the message text displayed when removing a product. That's not critical and could easily be ignored for a wp-admin version.

```js
import getPriceTierForUnits from 'calypso/my-sites/plans/jetpack-plans/get-price-tier-for-units';
```

This is a pure calculation function and could live anywhere. Moving it to `@automattic/calypso-products` seems sensible to me. It's only used by the next import to calculate Jetpack product tier subtitle text.

```js
import {
	getProductDisplayCost,
	getProductPriceTierList,
} from 'calypso/state/products-list/selectors';
```

As mentioned above, `getProductPriceTierList` is used to display a price tier for a Jetpack product. `getProductDisplayCost` is used to display the per-unit cost of P2 plan products. These both rely on a product list in Redux state and could be very difficult to migrate out of calypso. I'm not sure what to do about them.

michaeldcain commented:
I talked to @sirbrillig, and given the remaining work to pull out the dependencies for the wp-admin version, I was wondering if there's a version of this that could be launched to users earlier while the full-fledged popover cart is completed.

Unfortunately, since a user somewhat seamlessly transitions from Calypso to wp-admin in the gif above, we can't merge this current Calypso version until the behavior matches in wp-admin. Instead, we could launch a version to both wp-admin and Calypso that shows the cart icon (and the count) when items are present in the cart. This icon would link directly to Checkout in the "cart edit" mode (added in #51559).

It's probably worth handling that scaled back version in a separate issue. We can either keep this open for tracking, or could also merge this PR behind a feature flag while the wp-admin version is being worked on (to not have to continually rebase this). What are your thoughts, @sirbrillig and @jsnajdr?

jsnajdr commented:
Extracting certain widgets (Happychat, Inline Help) so that they can be embedded in wp-admin screens is something that I and @flootr are working on right now. The Notification panel can serve as a blueprint for this, because it's been independent like this for several years now. Let me have a look at what it takes to extract all the needed dependencies for Shopping Cart.

sirbrillig commented:
Yeah, all we should need in theory is the `@automattic/shopping-cart` package, some UI pieces to make a menubar item, and access to the REST API which should be available in wp-admin through wpcom-proxy-request.

jsnajdr commented:
> These both rely on a product list in Redux state and could be very difficult to migrate out of calypso.

The good news that the Redux product list is simply fetched and stored from a REST endpoint and doesn't have any other interactions with the state. And the `getProductPriceTierList` and `getProductDisplayCost` merely return some property of a given item in the products list. After unrolling all abstractions, the product list can be implemented as:
```js
function useProductList() {
  const [ list, setList ] = useState( {} );
  useEffect( () => {
    wpcom.req.get( '/products' ).then( list => setList( mapValues( list, ensureNumericCost ) ) );
  }, [] );
  return {
    getProductPriceTierList: ( slug ) => list[ slug ]?.price_tier_list ?? [],
    getProductDisplayCost: ( slug ) => list[ slug ]?.cost_display ?? null,
  }
}
```
The same hook, with the same API and behavior, can be implemented also using Redux, with `useDispatch( requestProductsList() )` and useSelector`.

The first implementation can be used in a wp-admin widget, the second one can be used in Calypso.

jsnajdr commented:
I see the following props that need to be injected from outside into the `MasterbarCart`, because their source is different in Calypso and wp-admin:
- `wpcom`: the REST API client. In Calypso, comes from `lib/wp`; in wp-admin, uses the `wpcom` package to create a client. [See how the Notifications widget creates it](https://github.com/Automattic/wp-calypso/blob/trunk/apps/notifications/src/standalone/auth-wrapper/index.jsx#L19) (in iframe proxy and OAuth variants) and injects as a `wpcom` prop.
- `currentUser`: in Calypso, comes from Redux; in wp-admin, we need to issue the `/me` REST request and pass the result. Used to determine logged-in/logged-out status, user flags and user currency code
- `selectedSite`: in Calypso, comes from Redux; in wp-admin, we need to figure the site ID somehow and then issue a REST request for `/sites/:site` to get the full object. Which site props does the Composite Checkout really use? If it's just ID and `slug`, our job is easier.
- `productsList`: see previous comment

jsnajdr commented:
> `getProductDisplayCost` is used to display the per-unit cost of P2 plan products.

Could this be replaced with the `product.product_cost_display` field that's already part of the Cart response? They both seem to show exactly the same info.

sirbrillig commented:
> These notice actions are used to construct callbacks like showErrorMessage passed as props to CheckoutProvider. It should be possible to replace the callback with another implementation, showing wp-admin notices.

These should no longer be necessary since https://github.com/Automattic/wp-calypso/pull/55809

sirbrillig commented:
> make the delete button a text link via new prop

I love this, and I actually think this may be better than the icon in checkout too. I've made a PR to explore that possibility, which would greatly simplify the changes here: https://github.com/Automattic/wp-calypso/pull/56045

sirbrillig commented:
> should we remove the popover carts from the domains and plans pages in this PR?

I removed the popover cart here for now. We can split that out if we like because it's hundreds of lines.

sirbrillig commented:
Some of the line item dependencies:

- https://github.com/Automattic/wp-calypso/pull/56176 moves getIntroductoryOfferIntervalDisplay to the `wpcom-checkout` package.
- https://github.com/Automattic/wp-calypso/pull/56177 moves getPriceTierForUnits to the `calypso-products` package.

sirbrillig commented:
Note to self: this appears to be where the Jetpack masterbar widgets currently are created: https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/modules/masterbar/masterbar/class-masterbar.php

jsnajdr commented:
> this appears to be where the Jetpack masterbar widgets currently are created

@sirbrillig I've been researching the same task (adding Help and Chat buttons to Masterbar) and figured out that in the wpcom codebase, for Simple sites, this is done very differently. There's a `wp-content/mu-plugins/notes/notifications-ui.php` for that. I'll ping you in a relevant Slack thread.

sirbrillig committed:
Add initial MasterbarCart app

sirbrillig commented:
> Could this be replaced with the product.product_cost_display field that's already part of the Cart response? They both seem to show exactly the same info.

Good question. It turns out they are not the same when the product has a `quantity` (eg: number of p2 plan user seats). However, we should be able to use `item_original_cost_display` for this... except that that calculation _also_ appears to be multiplied by `quantity` for some reason. I did some digging and I think that's a bug which I'm attempting to fix in D67069-code (if not, we can add it as a new property).

sirbrillig commented:
> As mentioned above, getProductPriceTierList is used to display a price tier for a Jetpack product... These both rely on a product list in Redux state

We could actually avoid needing the products list entirely because this data could easily be added to the shopping-cart endpoint. I've set that up in D67072-code.

a8ci18n commented on client/layout/masterbar/masterbar-cart.tsx:
```diff
@@ -0,0 +1,118 @@
+		} );
+	};
+	const onClose = () => setIsActive( false );
+	const tooltip = translate( 'View my Shopping Cart' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **34** times:
`translate( 'My Shopping Cart' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/masterbar-cart)

sirbrillig commented:
I've got an standalone app running the masterbar cart in https://github.com/Automattic/wp-calypso/pull/56369 

There's a pile of small dependency adjustments there that I'll need to make into separate PRs and then we'll need to see if it works in wp-admin. I'm wondering if the technique of embedding it in an iframe will be successful since the popup appears and disappears, but I guess we'll see. Right now I'm embedding the button and the popup in the same app but I wonder if the button needs to be native and just the popup contents in the iframe? 🤔

sirbrillig commented:
I rebased and bought it all the changes from my app branch. I think it's best to work on this in one place. The dependencies are very small now. The next step should be to try one of two things:

1. Create a WP plugin to embed this app directly in the WordPress masterbar.
2. Create a WP plugin to embed the app as an iframe in the WordPress masterbar.

There's advantages and disadvantages to both. Unlike [Inline Help](https://github.com/Automattic/wp-calypso/pull/55189), the masterbar cart should not need any API privileges apart from the current site's `/me/shopping-cart`, so using an iframe doesn't gain us as much as that PR and has some big costs, so I'd like to see if embedding the app code directly would work.

a8ci18n commented on client/layout/masterbar/masterbar-cart-button.tsx:
```diff
@@ -0,0 +1,64 @@
+		} );
+	};
+	const onClose = () => setIsActive( false );
+	const tooltip = translate( 'View my Shopping Cart' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **35** times:
`translate( 'My shopping cart' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/masterbar-cart)

sirbrillig commented:
I rebased and squashed a bunch of the commits together as this was getting rather large. I also moved the removal of the PopoverCart to [its own PR](https://github.com/Automattic/wp-calypso/pull/57520).

I'm discussing using the Jetpack API directly without an iframe in p9dueE-3KR-p2 but in the meantime I'll assume that the app created here will run inside an iframe.

sirbrillig committed:
Use MiniCart in apps/masterbar-cart

sirbrillig committed:
Remove masterbar styles from app imports

sirbrillig commented:
Updated the code to use a separate npm package, `@automattic/mini-cart` for the cart contents, then used that package in both calypso and the app.

sirbrillig committed:
Add emotion/react to masterbar-cart app deps

sirbrillig commented:
Tests are failing because of this error that doesn't make a lot of sense to me:

```
  Error: ➤ YN0000: │ wp-calypso@workspace:. STDOUT packages/mini-cart/src/mini-cart-line-items.tsx(6,8): error TS2307: Cannot find module '@automattic/wpcom-checkout' or its corresponding type declarations.
```

`'@automattic/wpcom-checkout'` is listed in the package dependencies for mini-cart. I wonder if it's trying to find it as a public package instead of looking in the monorepo?

sirbrillig committed:
Replace authWrapper with createClient in masterbar-cart app

This is the approach created in
https://github.com/Automattic/wp-calypso/pull/56958

sirbrillig commented:
The bundle size of the standalone app appears to be about 3.2 _megabytes_ which is... really huge for such a tiny display. I've tried to use the bundle size analyzer to find out what's taking up all that space but there doesn't seem to be any major offenders, just a collection of small ones, notably a bunch of `@emotion` packages, `@wordpress/data`, `wpcom.js`, `@automattic/calypso-products`, and `lodash-es`.

<img width="556" alt="Screen Shot 2021-11-03 at 8 11 23 PM" src="https://user-images.githubusercontent.com/2036909/140236073-44ab45d7-e5ed-481f-a041-776089c0cd53.png">

About a megabyte of that is `@automattic/wpcom-checkout`, which is primarily the code that displays all the different variations of line items we have. By displaying a much simplified view of each line item (without special sublabels, discounts, domains, email accounts, etc.) we can remove that dependency, but it's not ideal.

Beyond that, I'm not sure what else can be shaved off. `@emotion` is used to add styles to `composite-checkout` in a way that makes it easy to embed and theme CSS from an npm package. `lodash` is a dependency of `@wordpress/data`, which is used by `composite-checkout` to create a default registry for its children... it may be possible to remove that since it's not really used by the package itself. I'll experiment.

sirbrillig commented:
Removing `@wordpress/data` (and thereby `lodash`) from `composite-checkout` is quite effective. When combined with removing `wpcom-checkout` it brings the bundle size down to 1.9 megabytes. Still big but much more reasonable. I've started working on that in https://github.com/Automattic/wp-calypso/pull/57628

sirbrillig commented:
Extracted just the calypso version of the masterbar cart, using a monorepo package for the contents, into https://github.com/Automattic/wp-calypso/pull/58467. The wp-admin version can continue to be worked on here.

sirbrillig commented:
Now that the mini-cart is available in calypso, this can be rebased to be specific to the wp-admin masterbar cart.

sirbrillig committed:
Update masterbar-cart app dep versions

sirbrillig committed:
Update lockfile

sirbrillig committed:
Add webpack-bundle-analyzer to app deps

sirbrillig committed:
Add redux to app deps

sirbrillig committed:
Update lockfile again

sirbrillig committed:
Fix webpack config to load

sirbrillig committed:
Update lockfile again again

sirbrillig committed:
Add missing MiniCart props to app

sirbrillig commented:
I've put this on hold while the structure of using calypso apps in wp-admin is in heavy discussion. See the p2 posts here:

- p4TIVU-9XY-p2
- p9oQ9f-K6-p2
